### PR TITLE
Add API reference documentation generator

### PR DESCRIPTION
**Pull request by szemate**
See [`README.md`](https://github.com/autonity/docs.autonity.org/blob/api-doc-generator/README.md#generating-api-reference-documentation) and [`_apidocgen/README.md`](https://github.com/autonity/docs.autonity.org/blob/api-doc-generator/_apidocgen/README.md) for details.

TODO:
- [x] Finalize excluded contract functions
- [x] ~Add generated API docs for v0.14.0 (or v0.14.1?)~
- [x] Remove doc comments with the `@dev` NatSpec tag
- [x] Show state mutability
- [x] Remove input parameters & add 4-byte function selector
- [x] Add command for patching Autonity <0.15 to build NatSpec data files

### CODE REVIEW

**Comment by dweng0** [comment]
does the `.gitignore` file need python related files added to it?

**Comment by szemate** [comment]
> does the .gitignore file need python related files added to it?

There is a .gitignore file in the _apidocgen directory with the Python related files, I don't think we'll have Python code outside it

File: README.md [hunk 1]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+**TL/DR:** Symlink an [autonity](https://github.com/autonity/autonity) repository
+clone into the root of this repository and run [`apidocgen`](./_apidocgen/). The
+configuration is in [`config.toml`](./config.toml); the generated files are saved
+into `./contracts/<autonity-tag>/`.
+
+### Setup
+
+1. Clone the Autonity Go Client into an arbitrary directory:
+   ```sh
+   git clone git@github.com:autonity/autonity.git /path/to/autonity
```
Comment (szemate) [inline README.md]: I'm not sure whether we need such detailed instruction--what level of expertise do we expect the audience to have?

File: README.md [hunk 2]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+
+- Either clone the [Autonity Go Client](https://github.com/autonity/autonity) into the
+  root of this repository
+- Or add a symlink to an existing clone into this repository with `ln -s /path/to/autonity`
+
+### Adding API documentation for a new Autonity release
+
+1. If there are new contracts to document, add them to [`config.toml`](./config.toml).
+   The configuration is documented [here](./_apidocgen/README.md#configuration).
+1. Switch to the release tag in Autonity, e.g. for v0.14.0:
```
Comment (jay-cousins) [inline README.md]:
```suggestion
2. Switch to the release tag in Autonity, e.g. for v0.14.0:
```

File: README.md [hunk 3]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+   ```sh
+   cd autonity
+   git checkout v0.14.0
+   ```
+1. Compile the protocol contracts:
```
Comment (jay-cousins) [inline README.md]:
```suggestion
3. Compile the protocol contracts:
```

File: README.md [hunk 4]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+   ```sh
+   make contracts
+   ```
+1. Generate the API reference documentation:
```
Comment (jay-cousins) [inline README.md]:
```suggestion
4. Generate the API reference documentation:
```

File: README.md [hunk 5]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+   ```sh
+   cd -
+   apidocgen
+   ```
+1. Review and commit the created Markdown files.
```
Comment (jay-cousins) [inline README.md]:
```suggestion
5. Review and commit the created Markdown files.
```

File: README.md [hunk 6]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+
+### Developing contract documentation
+
+Autonity protocol contracts are documented via documentation comments in
+Solidity source code. To help with previewing documentation, a script is
+available that watches the source directory in Autonity and rebuilds the
+documentation when files change.
+
+1. Launch the monitoring script with `apidocmon`.
+1. Open the preview URL from the terminal in the browser.
```
Comment (jay-cousins) [inline README.md]:
```suggestion
2. Open the preview URL from the terminal in the browser.
```

File: README.md [hunk 7]
```diff
@@ -21,6 +21,60 @@ Usage:
...
+1. Open Solidity code from the symlinked Autonity repository clone in a
```
Comment (jay-cousins) [inline README.md]:
```suggestion
3. Open Solidity code from the symlinked Autonity repository clone in a
```

File: _apidocgen/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+## Usage
+
+The tool can be executed simply with `apidocgen` inside the devenv shell.
+
+A (symlink to a) clone of the [Autonity repository](https://github.com/autonity/autonity)
+is assumed to be in the working directory, or its path can be specified with the
+`--autonity` command line option.
+
+The configuration is read from a TOML file, defaults to `config.toml` in the
```
Comment (jay-cousins) [inline _apidocgen/README.md]:
```suggestion
The configuration is read from a TOML file. This defaults to `config.toml` in the
```

File: _apidocgen/README.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+working directory, or its path can be specified with the `--config` command line
+option.
+
+The `--watch` command line option launches an observer that detects contract code being
+modified and automatically recompiles the contracts and regenerates the Markdown
+documentation. Used together with `quarto perview`, it is possible to automatically
```
Comment (jay-cousins) [inline _apidocgen/README.md]:
```suggestion
documentation. Used together with `quarto preview`, it is possible to automatically
```

File: _apidocgen/compiler.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
+"""Wrapper for the Solidity compiler.
+
+Generates Solidity contract artefacts. To be used with the --watch option.
+
+It is needed because `make contracts` in Autonity takes ~20 seconds to complete
+therefore it isn't suitable for live previews. This module generates the artefacts
+in less than 1 second.
+
+Might need to be reworked later because it assumes that Nix's solc version matches
```
Comment (jay-cousins) [inline _apidocgen/compiler.py:9]: I think that's a sound assumption.

File: _apidocgen/contract.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+import re
+from enum import Enum
+from os import path
+from typing import TypeAlias
+
+from . import DEBUG
+from .paths import Paths
+from .markdown import MarkdownDocument
+
+NATSPEC_VERSION = 1
```
Comment (jay-cousins) [inline _apidocgen/contract.py:18]: I can't see the `kind` field?

https://docs.soliditylang.org/en/latest/natspec-format.html#documentation-output

>Note
>
>Starting Solidity version 0.6.11 the NatSpec output also contains a version and a kind field. Currently the version is set to 1 and kind must be one of user or dev. In the future it is possible that new versions will be introduced, deprecating older ones.

File: config.toml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# API reference documentation generator configuration
+# See: ./_apidocgen/README.md
+
```
Comment (jay-cousins) [inline config.toml]: I think this is a valid first iteration of config.toml exclusion 👍 🌟 ⭐

**Review by jay-cousins — CHANGES_REQUESTED** [review]
Looks good, some minor comments/questions

File: README.md (hunk 2 above)
Comment (szemate) [reply README.md]: From the [markdown reference](https://daringfireball.net/projects/markdown/syntax#list):
> It’s important to note that the actual numbers you use to mark the list have no effect on the HTML output Markdown produces. 

i.e. the list
```
1. one
2. two
3. three
```
produces the same output as
```
1. one
1. two
1. three
```
or
```
1. one
10. two
100. three
```
You can check it if you click the "rich diff" button in GitHub diff view.

File: _apidocgen/contract.py (hunk 11 above)
Comment (szemate) [reply _apidocgen/contract.py:18]: I assumed that the file extensions are correct, i.e. if it's `*.docuser` then `kind` is `user` and if it's `*.docdev` then `kind` is `dev`. Nevertheless I could add an assertion that checks it.

File: README.md (hunk 2 above)
Comment (jay-cousins) [reply README.md]: I know. I just wouldn't do it, hence the comment. But never mind, closing!

Comment (szemate) [reply README.md]: If you prefer having the right numbers, I wouldn't like to mess up the convention in the repo.

**Review by jay-cousins — APPROVED** [review]

**Review by aiman — APPROVED** [review]
Instead of `/config.toml` I think something like `/apidoc.toml`, `apiconfig.toml`, or `/docgen.toml` may be more self-descriptive. 

Otherwise LGTM

**Comment by szemate** [comment]
I think I've addressed all comments, @dweng0 could you please run the script and check whether the generated output is as expected?

**Comment by dweng0** [comment]
> I think I've addressed all comments, @dweng0 could you please run the script and check whether the generated output is as expected?

Off the back of this  @szemate  we tried running the scripts and ran into a problem. I know you are looking at it from your end. i'll be seeing if i can debug it from my end also.

For reference. I couldn't run `apidocgen` directly and had to manually run `python3 -m ./_apidocgen/apidocgen` which resulted in the following error:
``` sh
    117997:     __vdso_gettimeofdaysubprocess.CalledProcessError: Command '['git', 'describe', '--tags']' died with <Signals.SIGSEGV: 11>.
```

There was no other trace.

I've attempted to rule out Github by running the command that caused the error 

``` bash
git describe --tags
```

But that works as expected.

**Comment by dweng0** [comment]
@szemate  this working for me and has all the parts we are looking for. Here is an example from the generated script

**Pull request merged**